### PR TITLE
fix: repair dead discussion mentor-counting branch

### LINKED ISSUE #774: Discussion mentor counting is dead code (GraphQL + ignore_comment self-reference + dict-vs-object mismatch)
I noticed three layered defects while raising test coverage to 100%; together they make `enable_mentor_count` a no-op (or a crash) for discussions.

## Where

`most_active_mentors.count_comments_per_user`, the discussion branch at `most_active_mentors.py:115-130`, and `discussions.get_discussions` GraphQL query at `discussions.py:25-50`.

## What is wrong

1. **GraphQL fetches no comment author data.** The discussion query only requests `comments(first: 1) { nodes { createdAt } }`, so each comment node has no `user`/`author` field.
2. **Production would AttributeError if it ever got past the guard.** `most_active_mentors.py:118-122` does `comment.user` / `comment.user.login` / `comment.submitted_at` / `comment.ready_for_review_at` on what is actually a plain dict. Attribute access on dicts raises `AttributeError`.
3. **Self-reference in the ignore check.** Even if the data were present, `most_active_mentors.py:117-119` passes `comment.user` as both `issue_user` AND `comment_user` to `ignore_comment`, which evaluates `comment_user.login == issue_user.login` and always returns True, so every comment gets skipped.

## Why it has not been spotted

`enable_mentor_count` defaults to False, and the GraphQL query returns `comments(first: 1)` so `len(discussion["comments"]["nodes"]) > 0` is rarely true in a way that would surface the crash.

## What a proper fix needs

- Expand the GraphQL query to fetch comment authors (and the discussion author, for the ignore-self check) with proper pagination.
- Replace attribute access with dict access in the discussion branch of `count_comments_per_user`.
- Thread the discussion author through as `issue_user` instead of reusing `comment.user`.
- Add a real test that runs against a representative GraphQL discussion response and asserts non-empty `mentor_activity`.

## Why this is tracked separately

I bumped `--cov-fail-under` to 100 in a separate PR. To keep that PR scoped to coverage, I marked the dead block with `# pragma: no cover` and referenced this issue. Removing the pragma is part of the proper fix above.

### PR DESCRIPTION
**Pull request by kaizeenn**
## Description

Fixes #774.

Three layered defects made `enable_mentor_count` a complete no-op (or a crash) for GitHub Discussions:

### 1. GraphQL fetched no comment author data

`discussions.py` queried `comments(first: 1) { nodes { createdAt } }` — no `author` field, and only 1 comment. Expanded to `comments(first: 100)` with `author { login __typename }` on each comment node. Added `author { login __typename }` to the discussion node itself so the opener's login is available for the self-comment filter.

### 2. Attribute access on dict nodes would AttributeError

`most_active_mentors.py` used `comment.user.login`, `comment.submitted_at`, `comment.ready_for_review_at` on plain GraphQL dicts. Replaced with `comment.get('author') or {}` dict access throughout.

### 3. Self-reference in `ignore_comment`

The old code passed `comment.user` as **both** `issue_user` and `comment_user`, so the guard `comment_user.login == issue_user.login` always fired and every comment was skipped. Fixed by threading `discussion['author']['login']` as the discussion-author login and comparing directly, without going through `ignore_comment` (which expects github3 objects, not dicts).

### Structural fix

Moved the discussion block **outside** the `if issue:` guard so it actually runs when the call-site passes `issue=None, discussion=<dict>` (as `issue_metrics.py` does).

## Development notes

Changed files: `discussions.py`, `most_active_mentors.py`, `test_most_active_mentors.py`.

Added `TestCountCommentsDiscussions` with 7 cases:
- basic count
- author self-comment is ignored
- bot commenter is ignored
- `ignore_users` respected
- multiple distinct commenters
- empty comments list
- null author (deleted account) handled gracefully

`pytest test_most_active_mentors.py` — **11 passed** (4 baseline + 7 new). Commit signed off (DCO).

### CODE REVIEW

**Comment by jmeridth** [comment]
@kaizeenn looks like we have some python linting failing.

**Comment by kaizeenn** [comment]
Hi @jmeridth @zkoppert, I've formatted the code with black and all CI checks are now green. Ready for review!

**Comment by zkoppert** [comment]
Hey @kaizeenn! It looks like the copilot review bot raised some good points. Could you take a look at the fixes for those?

**Comment by jmeridth** [comment]
:robot: Now that the PyGithub migration (#789) has landed on main, this PR will need a rebase. The good news is that the actual logic in the PR is still correct and does not need to change.

Here is what to expect when rebasing:

- **`discussions.py`** applies cleanly. The GraphQL query changes do not overlap with anything from #789.
- **`most_active_mentors.py`** will conflict in the discussion branch block (lines 121-141). The resolution is straightforward since the PR replaces that same dead code block. The only contextual difference is that `ignore_comment` lost its `github3.users.User` type annotations and gained a `None` guard for ghost users, but that does not affect this PR since the discussion branch does its own inline dict-based filtering.
- **`test_most_active_mentors.py`** will conflict at the end of the file. The migration added two new test classes (`TestGhostUserHandling` and `TestMostActiveMentorsExtraBranches`) after the spot where this PR appends `TestCountCommentsDiscussions`. Just place the new class after the existing ones.

**Comment by jmeridth** [comment]
@kaizeenn I'm going to take over this PR. Your commits will stay. We need this to get merged so we can work on https://github.com/github-community-projects/issue-metrics/issues/790. Zero offense meant. Thank you for kicking this off and getting it 90% of the way there. :bow:

File: most_active_mentors.py [hunk 1]
```diff
@@ -112,27 +112,36 @@ def count_comments_per_user(
...
-                    continue
+    # The discussion branch: use dict access because GraphQL returns plain
+    # dicts (not github3 objects). Thread the discussion author as
+    # issue_user so we can filter out self-comments correctly.
+    if discussion and len(discussion["comments"]["nodes"]) > 0:
+        discussion_author_login = (
+            discussion.get("author") or {}
+        ).get("login", "")
+        for comment in discussion["comments"]["nodes"]:
+            comment_author = comment.get("author") or {}
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply most_active_mentors.py:138]: @copilot-pull-request-reviewer Good catch. The discussion branch now respects `max_comments_to_eval` using the same counter and break the issue-comment branch uses. Addressed in https://github.com/github-community-projects/issue-metrics/pull/776/commits/ac752fdd4daa2424b657dc53dbf2eef5854a25fc.

File: most_active_mentors.py [hunk 2]
```diff
@@ -112,27 +112,36 @@ def count_comments_per_user(
...
 
-                # increase the number of comments left by current user by 1
-                if comment.user.login in mentor_count:
-                    mentor_count[comment.user.login] += 1
-                else:
-                    mentor_count[comment.user.login] = 1
+            if comment_login in mentor_count:
+                mentor_count[comment_login] += 1
+            else:
+                mentor_count[comment_login] = 1
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply most_active_mentors.py:160]: @copilot-pull-request-reviewer Good catch. The discussion branch now caps per-user counts at `heavily_involved`, matching the issue-comment branch. Addressed in https://github.com/github-community-projects/issue-metrics/pull/776/commits/ac752fdd4daa2424b657dc53dbf2eef5854a25fc.

**Comment by jmeridth** [comment]
:robot: Picked this up to get it across the line.

- Rebased onto `main` now that the PyGithub migration (#789) has landed. `discussions.py` applied cleanly. `most_active_mentors.py` and the tests conflicted only where expected and were resolved without logic changes (the discussion branch moved outside the `if issue:` guard as intended).
- Addressed both Copilot findings: the discussion branch now respects `max_comments_to_eval` and caps per-user counts at `heavily_involved`, matching the issue-comment branch (ac752fd).
- Added edge-case tests (null `createdAt`, null discussion author) and documented the `discussion` parameter (6a36243).
- Filed #798 to track adding pagination to the 100-comment discussion fetch. It is harmless at the default `MAX_COMMENTS_EVAL` of 20, so it is documented inline here and left out of scope.

Tests: 21 passing with 100% coverage on the touched modules. black, mypy, and pylint (10.00/10) clean.

File: discussions.py [hunk 3]
```diff
@@ -31,9 +31,21 @@ def get_discussions(token: str, search_query: str, ghe: str):
...
-                        comments(first: 1) {
+                        author {
+                            login
+                            __typename
+                        }
+                        # Only the first 100 comments are fetched (no
+                        # pagination). MAX_COMMENTS_EVAL defaults to 20, so
+                        # this ceiling is not hit in practice; setting it above
+                        # 100 would silently cap discussion mentor counts.
+                        comments(first: 100) {
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply discussions.py:42]: @copilot-pull-request-reviewer Agreed, and this is a known limitation rather than an oversight. The 100-comment ceiling is documented inline directly above this query, and it is harmless at the default `MAX_COMMENTS_EVAL` of 20. Adding comment pagination is a larger change that is out of scope for this PR, which is focused on repairing the dead discussion mentor-counting branch. It is tracked as a follow-up in #798.

**Review by zkoppert — APPROVED** [review]
